numba/stencilparfor.py

```diff
@@ -0,0 +1,619 @@
+import numpy as np
+import numbers
+import copy
+
+def stencil():
+    pass
+
+@infer_global(stencil)
+class Stencil(AbstractTemplate):
+    def generic(self, args, kws):
+        assert not kws
+        return signature(types.none, *args)
+
+class StencilPass(object):
+    def __init__(self, func_ir, typemap, calltypes, array_analysis, typingctx):
+        self.func_ir = func_ir
+        self.typemap = typemap
+        self.calltypes = calltypes
+        self.array_analysis = array_analysis
+        self.typingctx = typingctx
+
+    def run(self):
+        from numba.stencil import StencilFunc
+        from numba import compiler
+
+        call_table, _ = get_call_table(self.func_ir.blocks)
+        stencil_calls = []
+        stencil_dict = {}
+        for call_varname, call_list in call_table.items():
+            if isinstance(call_list[0], StencilFunc):
+                stencil_calls.append(call_varname)
+                stencil_dict[call_varname] = call_list[0]
+        if not stencil_calls:
+            return  # return early if no stencil calls found
+
+        # find and transform stencil calls
+        for label, block in self.func_ir.blocks.items():
+            for i, stmt in enumerate(block.body):
+                if (isinstance(stmt, ir.Assign)
+                        and isinstance(stmt.value, ir.Expr)
+                        and stmt.value.op == 'call'
+                        and stmt.value.func.name in stencil_calls):
+                    kws = dict(stmt.value.kws)
+                    input_dict = {i: stmt.value.args[i] for i in
+                                                    range(len(stmt.value.args))}
+                    in_args = stmt.value.args
+                    arg_typemap = tuple(self.typemap[i.name] for i in
+                                                                stmt.value.args)
+                    for arg_type in arg_typemap:
+                        if isinstance(arg_type, types.BaseTuple):
+                            raise ValueError("Tuple parameters not supported for stencil kernels in parallel=True mode.")
+
+                    if 'out' in kws:
+                        out_arr = kws['out']
+                    else:
+                        out_arr = None
+
+                    sf = stencil_dict[stmt.value.func.name]
+                    stencil_blocks, rt, arg_to_arr_dict = get_stencil_blocks(sf,
+                            self.typingctx, arg_typemap,
+                            block.scope, block.loc, input_dict,
+                            self.typemap, self.calltypes)
+                    index_offsets = sf.options.get('index_offsets', None)
+                    gen_nodes = self._mk_stencil_parfor(label, in_args, out_arr,
+                            stencil_blocks, index_offsets, stmt.target, rt, sf,
+                            arg_to_arr_dict)
+                    block.body = block.body[:i] + gen_nodes + block.body[i+1:]
+                    return self.run()
```

Comment:
It doesn't seem the recursion is needed.  Am I missing something?

Reply:
Can replacing the block body make the loop state invalid? Or is the first `block.body` reference used once in the beginning and not anymore?

Reply:
My go-to for this problem is enumerating backwards.  How about:
for i, stmt in reversed(list(enumerate(block.body))): ?

Reply:
Tested this reversed, non-recursive approach and seems to work so it is checked-in.